Add support for user-supplied vertex shaders

andrewbaxter opened the pull request:
So the main things
* I changed `from_wgsl` to `new_wgsl` and made it buildery, there were a hundred ways I could have done this though.  New builder type? Mutable and use `&mut self` to modify? I have no idea the best choice here, but returning new objects means you have more flexibility chaining and storing.
* The examples I've seen have both the vertex and fragment shader in a single source file, so I made the `with_vertex` method just store the new entry point.  Should this actually take two different shader files?
* The example doesn't work for me (WIP), it shows white stretching from the center of the screen in a couple directions. I think some sort of base matrices from `DrawUniforms` are lost when using the user-supplied stuff, but I couldn't figure out what they were (projection? or something else?).

I'm not sure what to do for the third point.  Should the user uniforms implement some interface to merge it with the base `DrawUniforms` data?  Or should methods be provided to get the data that would have gone into the original draw shader?

I think it would be useful to document somewhere "the default vertex shader gets this data and does these things, if you supply your own shader you can (use them with these names|get the data to include yourself by calling these methods)".

closes #1113

jazzfool commented:
In your example's vertex shader, you're trying to access your custom shader struct from ggez's reserved DrawUniform binding slot.

This is ggez's default draw uniforms bound to set/group 0, binding 0 (see [the default draw shader](https://github.com/ggez/ggez/blob/cbb3d202f512dc79b484829c122cfa677423e8c7/src/graphics/shader/draw.wgsl#L7)):
```rust
struct DrawUniforms {
    color: vec4<f32>,
    src_rect: vec4<f32>,
    transform: mat4x4<f32>,
}
```

Your actual custom uniforms are bound to set 3, binding 0.
You're basically reading off a nonsensical matrix when accessing `uniforms.transform`.

So all in all:
```rust
struct MyCustomDrawUniforms {
    transform: mat4x4<f32>,
    rotation: mat4x4<f32>,
}

struct GgezDrawUniforms {
    color: vec4<f32>,
    src_rect: vec4<f32>,
    transform: mat4x4<f32>,
}

@group(0) @binding(0)
var<uniform> uniforms: GgezDrawUniforms; // << not MyCustomDrawUniforms!

@group(3) @binding(0)
var<uniform> my_uniforms: MyCustomDrawUniforms;

@vertex
fn vs_main(
    @location(0) position: vec2<f32>,
) -> @builtin(position) vec4<f32> {
    return my_uniforms.transform * vec4<f32>(position, 0.0, 1.0);
}

@fragment
fn fs_main() -> @location(0) vec4<f32> {
    return vec4<f32>(1.0, 1.0, 1.0, 1.0);
}
```

One thing you'll want to do in the impl itself is ensure that everything in the user shader param bind groups are vertex visible, [here](https://github.com/ggez/ggez/blob/devel/src/graphics/shader.rs#L138). Possibly could make them conditionally vertex visible based on if a custom vertex shader was provided - I think this is desirable as the descriptor pointers in the SGPR aren't shared between stages, so chucking that vertex visibility flag on *is not free*, and it would also likely cause sub-optimal syncing within WGPU.

BTW, it would be ideal to pre-multiply `translation` and `rotation`, just to display best practice.

PSteinhaus commented:
Thanks for giving this a try!

The reason (or a reason) why your example doesn't work should be that you're binding `DrawUniforms` to group 0, while that's reserved for the ggez uniforms. group 4 is reserved for custom uniforms:
https://github.com/ggez/ggez/blob/28964dea0e0e7ae86c47e7020ca76bffb7533491/src/graphics/shader.rs#L80-L96

> * I changed `from_wgsl` to `new_wgsl` and made it buildery, there were a hundred ways I could have done this though.  New builder type? Mutable and use `&mut self` to modify? I have no idea the best choice here, but returning new objects means you have more flexibility chaining and storing.

Yes, new builder type and I'd say we use `self` to modify, so we can chain calls like this:
```rust
let test_shader: Shader = ShaderBuilder::new_wgsl().fragment_path(path, name).vertex_source(source, another_name).build();
```


>* The examples I've seen have both the vertex and fragment shader in a single source file, so I made the `with_vertex` method just store the new entry point.  Should this actually take two different shader files?

I'd say we want the following methods on our builder for that:
* `fragment_path` and `fragment_source` for creating a ShaderModule to store under `fragment` as we had it
* `vertex_path` and `vertex_source` for creating a ShaderModule to store under a new `Shader` field `vertex`
* `combined_path` and `combined_source` for creating a ShaderModule to store in `fragment` with a clone in `vertex`

The `Shader` created has `fragment` and `vertex` as fields of `Option<ArcShaderModule>`.
In `update_pipeline` we check if `vertex` is `Some(module)` and if not we use the vertex shader modules we currently use (i.e. default to this:)
https://github.com/ggez/ggez/blob/28964dea0e0e7ae86c47e7020ca76bffb7533491/src/graphics/internal_canvas.rs#L574-L584
For the fragment shader we do something similar and check whether `fragment` is `Some(module)` and if not we use `self.draw_sm.clone()` or `self.text_sm.clone()` depending on the `ShaderType`.


> I'm not sure what to do for the third point. Should the user uniforms implement some interface to merge it with the base `DrawUniforms` data? Or should methods be provided to get the data that would have gone into the original draw shader?

The user uniforms can be anything they like. They just need to use group 4 not 0 ^^

> I think it would be useful to document somewhere "the default vertex shader gets this data and does these things, if you supply your own shader you can (use them with these names|get the data to include yourself by calling these methods)".

Hmm. Maybe we could document/comment the shader files used in ggez better?

PSteinhaus commented:
One thing that we're not sure about is maybe hard-coding the stage names to `vs_main` and `fs_main` so that we can lose that `fs_entry`.
Only downside would be that users would have to follow this convention, but we can mention it in the `ShaderBuilder` docs, so it should be ok, I'd say.

andrewbaxter commented:
> Your actual custom uniforms are bound to set 3, binding 0.

Ah, thanks! Trying that now.

> I'd say we want the following methods on our builder for that:

Okay, so I guess this should also support a vertex shader with no fragment shader (and, maybe for development purposes, one with neither vertex nor fragment bits).

@PSteinhaus 
> fragment_path

Sorry, I meant to ask about this. Should this be a resource path instead of the shader file contents directly? The current `from_wgsl` takes the contents so I was following that convention. (I'll assume the former and start on changes here)

> so it should be ok, I'd say.

It sounds good to me, so I'll go ahead with that unless it was still pending concensus.

Thanks for all the help!

PSteinhaus commented:
> Okay, so I guess this should also support a vertex shader with no fragment shader (and, maybe for development purposes, one with neither vertex nor fragment bits).

Yes, absolutely. (Though an "empty" Shader is a bit ridiculous, but using two `Option`s it's just a feature that comes naturally, so why fight against it. Who knows, maybe it'll really be somewhat useful to someone in the developement process, as you said.)

> > fragment_path
> 
> Sorry, I meant to ask about this. Should this be a resource path instead of the shader file contents directly? The current `from_wgsl` takes the contents so I was following that convention. (I'll assume the former and start on changes here)

Yes, this variant should take a resource path, similar to what is possible with `Image::from_path` vs. `Image::from_pixels`.

> It sounds good to me, so I'll go ahead with that unless it was still pending concensus.

Yes, please do :)

> Thanks for all the help! I sincerely help this is a net benefit after all the assistance 😅

It is ^^
I wouldn't have been able to make these changes myself, as I'm currently busy at the university and will have to work this evening too.

PSteinhaus commented:
Thanks for your work! (And thanks for handling #1116 while you were at it ;) )
As I said I'm sadly busy this evening, but I'll try to check it tomorrow :)

andrewbaxter commented:
Sorry, I guess I needed to do `cargo clippy --examples` to get clippy to check everything. :crossed_fingers:

PSteinhaus commented on examples/vertex_shader.rs:
```diff
@@ -0,0 +1,93 @@
+        path::PathBuf::from("./resources")
+    };
+
+    let cb = ggez::ContextBuilder::new("colorspace", "ggez").add_resource_path(resource_dir);
```
This should say something different than "colorspace" probably ^^

PSteinhaus commented on src/graphics/shader.rs:
````diff
@@ -14,8 +18,133 @@ use super::{
+/// ```
+/// if the fragment shader is left unspecified.
+///
+/// Produce a Shader using ShaderBuilder.
````
A link to `ShaderBuilder` on the word would be nice here :)

PSteinhaus commented on src/graphics/shader.rs:
```diff
@@ -14,8 +18,133 @@ use super::{
+
+    /// Create a Shader from the builder.
+    pub fn build(self, gfx: &impl Has<GraphicsContext>) -> GameResult<Shader> {
+        let gfx = gfx.retrieve();
+        let load = |s: &str| {
+            Some(ArcShaderModule::new(gfx.wgpu.device.create_shader_module(
+                wgpu::ShaderModuleDescriptor {
+                    label: None,
+                    source: wgpu::ShaderSource::Wgsl(s.into()),
+                },
+            )))
+        };
+        let load_resource = |path: &str| -> GameResult<Option<ArcShaderModule>> {
+            let mut encoded = Vec::new();
+            _ = gfx.fs.open(path)?.read_to_end(&mut encoded)?;
+            Ok(load(
+                &String::from_utf8(encoded).map_err(GameError::ShaderEncodingError)?,
+            ))
+        };
+        let load_any = |source| -> GameResult<Option<ArcShaderModule>> {
+            Ok(match source {
+                ShaderSource::Code(source) => load(source),
+                ShaderSource::Path(source) => load_resource(source)?,
+                ShaderSource::None => None,
+            })
+        };
+        Ok(if self.vs == self.fs {
+            let module = load_any(self.vs)?;
+            Shader {
+                vs_module: module.clone(),
+                fs_module: module,
+            }
+        } else {
+            Shader {
+                vs_module: load_any(self.vs)?,
+                fs_module: load_any(self.fs)?,
+            }
+        })
+    }
+}
```
This function I like ^^

PSteinhaus commented on src/graphics/shader.rs:
```diff
@@ -53,33 +180,15 @@ use wgpu::util::DeviceExt;
 /// Parameters that can be passed to a custom shader, including uniforms, images, and samplers.
 ///
-/// These parameters are bound to group 4. With WGSL, for example,
+/// These parameters are bound to group 3. With WGSL, for example,
```
A bit below the `@group(4)`s need to be changed as well

andrewbaxter commented:
> This should be done as well.

It sounds like this is a design problem ATM? I don't have any wisdom here, I'm happy to implement it if someone has a concrete proposal.

PSteinhaus commented:
> > This should be done as well.
> 
> It sounds like this is a design problem ATM? I don't have any wisdom here, I'm happy to implement it if someone has a concrete proposal.
> 
> Edit: Is this as simple as adding another parameter, `make_vertex_visible`?
> 
> Edit: I guess this should go full-builder too.

Hmm. Simply adding a `make_vertex_visible` was the first thing I though of too.
How would you solve it with a builder?

Another idea might be to just hand the shader that we want to create to `ShaderParams::new` so that we can check internally whether the flag needs to be set. This would mean users wouldn't have to worry about `make_vertex_visible`, but it would also rob them of a bit control.

andrewbaxter commented:
Ah no, I was just thinking the amount of params is growing and not all of them are always relevant (ex, all examples use empty images/samplers).  I pushed a builder, happy to revert though.

That would work too... is it possible that there could be images/samplers in a shader with a vertex shader but they're not used by the vertex shader so we wouldn't want to make them visible?

andrewbaxter commented:
I think I've addressed all the above comments.

nobbele commented:
Documentation for `Shader` needs to be updated

andrewbaxter commented:
Ah... unfortunately I'm not familiar with shaders either (or graphics programming in general).  Should I combine those into one method?

I updated the doc string, I think it was just the `fragment` thing in the first line but I also changed some other terminology since it was bothering me, I think "shader" and "module" is at least more similar to the internal naming.

nobbele commented:
I meant the documentations referring to `ShaderParams`

I am thinking if we should perhaps have `.images(&[(my_image, my_sampler)])`

andrewbaxter commented:
Ah okay, I fixed the doc example code that was using `ShaderParams::new`.

I went ahead with the images change as well.

nobbele commented on src/graphics/shader.rs:
````diff
@@ -53,33 +180,15 @@ use wgpu::util::DeviceExt;
-/// These parameters are bound to group 4. With WGSL, for example,
+/// These parameters are bound to group 3. With WGSL, for example,
 /// ```rust,ignore
 /// ggez::graphics::ShaderParams::new(ctx, &my_uniforms, &[&image1, &image2], &[sampler1])
````
Outdated docs here, should be the only one left

andrewbaxter commented on src/graphics/canvas.rs:
```diff
@@ -563,29 +563,22 @@ struct DrawCommand {
     pub shader: Shader,
-    pub text_shader: Shader,
     pub mesh: Mesh,
     pub image: Image,
 }
 
 impl DefaultResources {
     fn new(gfx: &GraphicsContext) -> Self {
         let shader = Shader {
-            fragment: gfx.draw_shader.clone(),
-            fs_entry: "fs_main".into(),
-        };
-
-        let text_shader = Shader {
-            fragment: gfx.text_shader.clone(),
-            fs_entry: "fs_main".into(),
+            fs_module: None,
+            vs_module: None,
         };
 
         let mesh = gfx.rect_mesh.clone();
         let image = gfx.white_image.clone();
 
         DefaultResources {
             shader,
-            text_shader,
```
Yeah... I was thinking of deleting it and using a `Shader {}` directly everwhere but I really don't know what would be clearest here.

PSteinhaus replied:
I'm not sure about the best way here either, but I'd say we delete the shader from this defaults struct.

I think it would be best to specify such defaults (sampler is set as linear clamp explicitly everywhere as well, so it's a similar case) in pub static functions `default_sampler` and `default_shader`, `default_text_shader`, so we still have the defaults of everything in one place instead of duplicating them.

andrewbaxter commented:
I found two more references to bind group numbers in `canvas.rs`.  The one that said 4 I changed to 3, but for text it says bind group 3 - is that correct? AFAICT from the code the text and non-text shaders used the same bind group numbers (?)

https://github.com/ggez/ggez/blob/master/src/graphics/canvas.rs#L193

PSteinhaus commented:
> I found two more references to bind group numbers in `canvas.rs`. The one that said 4 I changed to 3, but for text it says bind group 3 - is that correct? AFAICT from the code the text and non-text shaders used the same bind group numbers (?)
> 
> https://github.com/ggez/ggez/blob/master/src/graphics/canvas.rs#L193

Hahh... thanks. That would have gone through and probably have become annoying somewhere down the road.
This was last changed here: https://github.com/ggez/ggez/commit/cbb3d202f512dc79b484829c122cfa677423e8c7

As all shaders use group 3 in `update_pipeline` 3 is correct for the text shader... weirdly.

jazzfool commented on src/graphics/shader.rs:
```diff
@@ -53,77 +180,59 @@ use wgpu::util::DeviceExt;
-    pub(crate) bind_group: ArcBindGroup,
-    _marker: PhantomData<Uniforms>,
-}
+    /// Provides images to the shaders.
+    ///
+    /// # Arguments
+    ///
+    /// * `vs_visible` - If the images should also be visible to the vertex shader, rather
+    ///    than just the fragment shader.
+    pub fn images(self, images: &'a [(&'a Image, Sampler)], vs_visible: bool) -> Self {
+        ShaderParamsBuilder {
+            uniforms: self.uniforms,
+            images,
+            images_vs_visible: vs_visible,
+        }
+    }
```
Ah, there's been a misunderstanding what sampler means in relation to images. You should *not* have 1 sampler to 1 image, so they shouldn't be bundled at all. It would be expected you reuse the same sampler in the shader for sampling N images. The most common use case is to have 1 sampler and N (>1) images. Of course you can have more than 1 sampler if you wish to sample differently, e.g., in the same shader you want to sample both linearly and nearest (which is quite a rare thing to do). Regardless, they should be taken separately and stored separately as you may have a different number of each -- however, if this at all helps with deciding API ergonomics, when you want to make the images vertex-visible, you almost certainly also want to make the samplers vertex-visible.

andrewbaxter replied:
Ah thanks! Okay I separated the lists again (I think that should be sufficient).

PSteinhaus replied:
Hmm. On one hand it's more straightforward to nudge/remind the user to hand over samplers (at least one) as well when handing over images. On the other hand we already have a builder and we could easily split this up into seperate functions for handing over images and samplers each (the concept of users having to hand over slices of them is probably worth keeping, as it makes it easier for users to see and control the order of samplers/images which they need to know while writing the shader).

But in the end I too think that this is sufficient.

nobbele replied:
could we pass the canvas's current sampler by default perhaps?

PSteinhaus replied:
Nah, let's better not hide some implicit magic here. Explicit is clearer and not too ugly imo.

andrewbaxter replied:
I'll push a separate branch if that's alright.

PSteinhaus commented:
Ok, I think we're ready to merge. Any objections?

andrewbaxter commented:
:tada: Thanks for the help and reviews!
